Add initial cut of a CSS style guide for Hypothesis

robertknight opened the pull request:
Based on lessons from the past few months of working with the Hypothesis
codebase and the guides linked herein, add an initial version of a
modern style guide for Hypothesis projects.

robertknight committed:
Add initial cut of a CSS style guide for Hypothesis

Based on lessons from the past few months of working with the Hypothesis
codebase and the guides linked herein, add an initial version of a
modern style guide for Hypothesis projects.

robertknight commented:
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;Snl4VEQ1&#x2F;398-write-a-css-style-guide-for-the-front-end"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> Write a CSS style guide for the front end</a>

robertknight commented on docs/css-style-guide.md:
```diff
@@ -0,0 +1,92 @@
+
+## Browser Support
+
+ * Hypothesis projects should be functional on Internet Explorer 10 and above. CSS features that are only available in newer browsers can still be used but the site should behave gracefully in older browsers
```
- [x] ~~Perhaps add a mention for using [@supports](https://developer.mozilla.org/en-US/docs/Web/CSS/@supports) here?~~

nickstenning replied:
A styleguide should describe real, tested practice. We don't use `@supports` anywhere in our code, so let's not codify if or how we use it in a styleguide until we understand it in practice.

nickstenning commented on docs/css-style-guide.md:
````diff
@@ -0,0 +1,92 @@
+```
+
+ * A 'base' partial which imports all variables and mixins used by other components
+ * A 'variables' partial which defines the colors, typography, grid spacing, animation durations and other foundational elements of the design system
````
Variables are in the base partial.

nickstenning commented on docs/css-style-guide.md:
```diff
@@ -0,0 +1,92 @@
+Hypothesis CSS Style Guide
+==========================
+
+Hypothesis' CSS generally follows the principles of [Trello's CSS Style Guide](https://github.com/trello/trellisheets)
```
I wouldn't start with a link that sends people off to a different CSS styleguide. By all means acknowledge the heritage at the end, but our styleguide should be (largely) self-contained, and not require or suggest that people read other folks' styleguides -- which, after all, will very likely change over time.

robertknight replied:
Agreed.

nickstenning commented on docs/css-style-guide.md:
```diff
@@ -0,0 +1,92 @@
+## Selectors
+
+ * Use simple non-nested class selectors for the majority of styling
+ * Descendant selectors should be avoided. They [are slow](https://developer.mozilla.org/en-US/docs/Web/Guide/CSS/Writing_efficient_CSS#Avoid_the_descendant_selector.21) and more importantly, make it easy to create unintended results due to naming clashes or [specificity conflicts](http://css.maxdesign.com.au/selectutorial/advanced_conflict.htm).
```
It seems to me like you're saying the same thing twice. "Nested selectors" == "descendant selectors", no? If so, we don't need two separate admonitions against them. The second is clearer and gives reasons.

robertknight replied:
Yes. The reason I originally included both is because in SASS you can write a descendant selector either directly or via nesting. Looking at styling issues I've fixed in the past, I don't think the hazards of nesting (which visually looks nice in SASS) was obvious. Perhaps best to spell this out with an example?

nickstenning replied:
Examples are almost always the best way to explain this stuff, IMO.

nickstenning commented on docs/css-style-guide.md:
```diff
@@ -0,0 +1,92 @@
+
+## Selectors
+
+ * Use simple non-nested class selectors for the majority of styling
```
Phrases like "for the majority of styling" are dangerous in styleguides. What is the minority for which deviation is acceptable?

seanh replied:
Doesn't linking to the Trello CSS guide and/or saying that we use BEM naming style cover not using nested / descendant selectors anyway?

nickstenning commented on docs/css-style-guide.md:
```diff
@@ -0,0 +1,92 @@
+## Layout
+
+ * Most layout should be done with block, inline or flexbox layout
+ * Floats should generally be avoided in favor of flexbox. This avoids the need for clearfix and other hacks. Note that flexbox [may not be appropriate](https://hyp.is/AVKcwo8BvTW_3w8LypJ5/jakearchibald.com/2014/dont-use-flexbox-for-page-layout/) for overall page layout
```
Again, I feel like this is saying the same thing twice. The second explanation is clear: "layout within components should usually be done with flexbox rather than with floats" and provides a rationale. It also explains which situations flexbox isn't appropriate in.

So let's just cut the first line?

nickstenning commented on docs/css-style-guide.md:
```diff
@@ -0,0 +1,92 @@
+
+## Colors and Typography
+
+ * Components should use colors and font sizes defined in the 'variables' partial (or one of its include) rather than hard-coding colors
```
Ditto -- probably meant to refer to the base partial.

nickstenning commented on docs/css-style-guide.md:
```diff
@@ -0,0 +1,92 @@
+
+ * Use variables and mixins
+ * [Avoid @extends](https://www.sitepoint.com/avoid-sass-extend/). Use mixins instead
+ * Use nesting sparingly
```
Might be worth clarifying what this means, because "nesting" could refer to several things. "Use nesting of rules sparingly -- overly nested SCSS can impede understanding."

nickstenning commented on docs/css-style-guide.md:
````diff
@@ -0,0 +1,92 @@
+ .component__element--modifier { ... }
+ ```
+
+ Where 'component' is a re-usable building block (eg. an input field),
````
"input field" seems like a weak example to me -- perhaps "date picker" or "dropdown menu"?

robertknight replied:
Perhaps - but components are not just for big elaborate pieces of UI, also for simple things like buttons. One trait of good libraries, conventions or tools for building pieces of UI is that it should be 'cheap' to extract out re-usable building block and compose them into others.

nickstenning commented on docs/css-style-guide.md:
```diff
@@ -0,0 +1,92 @@
+ '--modifier' represents some static variation of a component (eg.
+ `btn--danger` for a button which performs some destructive action)
+ * Use `is-$state` names for state classes (eg. `is-active`) that are dynamically added or removed from JS code
+ * Use names with a `js-` prefix for classes that are used by JS to get a reference to particular DOM elements. These classes should not appear in stylesheets
```
This is specifically not a rule about how to write CSS, so maybe move this to the bottom and say instead "Classes with a `js-` prefix are used by JS to...."?

seanh commented on docs/css-style-guide.md:
````diff
@@ -0,0 +1,92 @@
+
+ * Use [BEM-style](http://csswizardry.com/2013/01/mindbemding-getting-your-head-round-bem-syntax/) naming for components:
+
+ ```js
````
css

seanh commented on docs/css-style-guide.md:
```diff
@@ -0,0 +1,92 @@
+for writing maintainable styling. In particular:
+
+ * Use simple, non-nested class selectors where possible
+ * Organize CSS into components. Unlike Trello, we use BEM-style naming (see below) and do not indent selectors
```
What's the motivation for using BEM-style naming instead of just using Trello's style?

Quite a lot of the Trello CSS guide is describing their naming style / the whole thing is sort of centered around the naming style, I think it makes it a little difficult to follow the Trello style but use a different naming style.

Might be worth it if there's a significant advantage?

robertknight replied:
BEM-styling naming has the slight advantage that you can parse a class name into independently re-usable block/component and elements which only make sense within their containing block by looking just at the name.

eg. Looking at `.tab-bar__tab--dark` you can tell that "tab-bar" is a re-usable thing, 'tab' is a sub-element of it that you wouldn't use outside of a tab bar.

I've moved the Trello reference just to the Further Reading section now so it is clear that it provides guidance rather than a template for this guide.

seanh commented on docs/css-style-guide.md:
```diff
@@ -0,0 +1,92 @@
+ * Be careful with CSS preprocessor features. Unlike Trello, we use SASS but the principle
+   of using 'vanilla' CSS and being careful with preprocessor features still applies.
+   See the note about `@extends` below
+ * Use naming conventions to indicate which classes are purely presentational and which are referenced in code
```
This is referring to the use of `js-*` classes for JavaScript to find particular elements, and `is-*` classes for state that's added and removed by JavaScript? If so I think this sentence would be clearer if concrete, something like: use `js-*` classes to enable JavaScript code to find particular DOM elements, and `is-*` classes for states that are added and removed by JavaScript." And then maybe link to the section(s) that explain this in more detail.

robertknight replied:
> This is referring to the use of js-\* classes for JavaScript to find particular elements,
> and is-\* classes for state that's added and removed by JavaScript?

That is correct. I've updated the documentation to refer to the `js-` classes as handle or reference/"ref" classes.

seanh commented on docs/css-style-guide.md:
```diff
@@ -0,0 +1,92 @@
+
+The SASS for Hypothesis projects is organized into several parts:
+
+ * A 'root' file which imports all of the components used in the site, structured like this:
```
Doesn't this mean that _all_ of our CSS is downloaded for every page?

nickstenning replied:
There may be more than one "root" file. c.f. https://github.com/hypothesis/h/tree/master/h/static/styles

robertknight replied:
> Doesn't this mean that all of our CSS is downloaded for every page?

As Nick says, there can be more than one root file. eg. We don't serve the CSS for our admin pages to ordinary visitors to the site. This _does_ still mean that typically you would download more CSS than you need for any given page load, but it would save fetching another bundle when navigating to a different page.

This is a classic trade-off with bundling assets together. With HTTP 2 one can generally afford to use smaller, more granular bundles because the overhead of an individual request is lower and there isn't the same limit on the number of concurrent requests to a particular domain.

seanh commented on docs/css-style-guide.md:
```diff
@@ -0,0 +1,92 @@
+// Components
+@import 'partials/input-field';
+@import 'partials/fancy-modal';
+@import 'partials/a-widget';
```
How are we going to keep the HTML, JavaScript and CSS code for a component together? A component probably isn't just CSS, right? Wondering if we need something like:

```
a-widget/
  a-widget.html.jinja2
  a-widget.js
  a-widget.sass
```

robertknight replied:
Yes, that's one approach that we'll probably look at. Moving from the layout discussed here to this layout would just involve moving some files though.

robertknight commented on docs/css-style-guide.md:
```diff
@@ -0,0 +1,112 @@
+Hypothesis' CSS generally follows the principles of [Trello's CSS Style Guide](https://github.com/trello/trellisheets)
+for writing maintainable styling. In particular:
+
+ * CSS rules are organized into several three categories:
```
Note: I'm still re-working this section.

robertknight commented:
OK, thank-you for the detailed feedback on the first round @nickstenning and @seanh . I think I've addressed all of the original comments now:

Changes since the first iteration:
- The guide is now written to be understandable on its own without direct reference to any other guides
- Removed some duplication and added examples
- Added an 'Organization' section at the start of the guide
- Re-phrased all the bullet-points to use the imperative mood consistently. I'm not completely sure whether this is the right choice but it is at least consistent

nickstenning commented:
LGTM! Nicely done. :cake:
